### PR TITLE
Media file consistency module

### PR DESCRIPTION
**Pull request by WakeflyCBass**
This module adds three checks to the system:
1) Media Library records not associated with any file in the filesystem.
- This iterates over the media library and attempts to access the physical file for each. It then displays each record that was not found. This way the user can go find the missing file and re-add it if necessary.

2) Form Attachments records not associated with any file in the filesystem.
- This iterates over each bizform and finds any user-submitted attachments submitted. If Kentico believes one should exist, but the file is not in the file system, that record is returned and reported. This allows the owner to find any missing attachments that the user expects to have been submitted.

3) Form Attachments files not associated with any Form record in Kentico
- This iterates over each physical file in the Form Attachments folder, and then compares that list to the various forms' files. If no matching Form Attachment record is found, that file is reported. This allows the administrator to delete any file attachments that are no longer being tracked by Kentico.

It should properly work for all four supported Kentico cases of location storage: "C:/", "C:\", "\\server\" and "~/".

It does not support mixed-use folders: "c:/folder\subfolder/".  It also assumes that the KInspector package has file access to any Kentico attachment or media library item via the same pathing structure that the website uses. 
A future enhancement of this feature could use the Kentico public urls for the media library instead of attempting to access the files directly. This might allow us to run these tests without caring about the base folder structure - I'm not sure if there are edge cases where doing so would not return proper results.  
I don't believe there is a public URL for bizform attachments, so that case still needs to know the physical file structure.

### CODE REVIEW

**Comment by ChristopherJennings** [comment]
Hi @WakeflyCBass,

I'll take a look at this tomorrow and either pull it in or submit some feedback.

File: KInspector.Modules/Modules/Content/DBFileConsistencyModule.cs [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+			/// <param name="baseUri">The virtual, server, or physical path</param>
+			/// <returns></returns>
+			public static FileInfo[] GetFiles(string baseUri, bool recursive, IInstanceInfo info)
+			{
+				if(baseUri == null)
+				{
+					throw new ArgumentNullException($"baseUri value is null");
+				}
+
+				if(baseUri.StartsWith("~/"))
```
Comment (ChristopherJennings) [inline KInspector.Modules/Modules/Content/DBFileConsistencyModule.cs]: Recommend adding support for a single backslash as well since that is supported as well.

**Comment by WakeflyCBass** [comment]
Single forward-slash or backslash (which I believe is equivalent to relative path with no ~) – is supported?
https://docs.kentico.com/display/K8/Settings+-+Media
Doesn’t look like it, here.

**Comment by ChristopherJennings** [comment]
The documentation doesn't list "\SomeFolder" as supported explicitly, but it is basically a normal Windows path (relative to the application root). By luck, the first project I tested this against had a custom path set for form files set to "\UploadedFiles\SomeFolder". Perhaps it's best to ignore this edge case though and just leave it as only what's listed in the documentation.

File: KInspector.Modules/Modules/Content/DBFileConsistencyModule.cs [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+				{
+					//If we're using a relative path, it's relative to the "CMS" folder, not the Kentico base folder.
+					var absPath = Combine(info.Directory.FullName, "CMS", baseUri.Substring(2));
```
Comment (ChristopherJennings) [inline KInspector.Modules/Modules/Content/DBFileConsistencyModule.cs]: It is expected that the instance info will contain the path to the CMS directory, not the parent directory. It is therefore not necessary to inject the CMS folder into the path. Especially since this will not exist in versions older than 8 (hence the expectation that the directory is the CMS directory)

**Comment by WakeflyCBass** [comment]
I believe my documentation does specifically call out the three folder structures, and say that those are the only supported ones. If not, I can add it.

At the very least it shouldn’t be too hard to edit support in later, in any case. I wish that there was a unified URI/file handling .net method that actually handled edge scenarios like forward/back slashes, and could take a parameter of a base folder to base relative paths on.

**Comment by ChristopherJennings** [comment]
I've started the refactoring and submitted an example as a pull request to your copy. Please take a look and let me know what you think. As far as the directory support issue, I say it's fine as-is. If we want to add support for undocumented, but supported syntaxes, let's do that in the future.

**Comment by WakeflyCBass** [comment]
Alright, I've gone through and done some refactoring on my end - I also want to add some logic for making sure the bizform is actually configured to use the filesystem, or else to ignore that piece.

**Comment by WakeflyCBass** [comment]
Alright, I've gone and done some refactoring, thanks for the example. How are you feeling about the code now?

**Comment by ChristopherJennings** [comment]
I'm sorry, I must have missed this notification. I'll review this tomorrow and get back to you.

**Comment by ChristopherJennings** [comment]
Looks good! Thanks for all your your hard work.
